src/Unleash/Scheduling/SystemTimerScheduledTaskManager.cs

```diff
@@ -80,34 +83,47 @@ async void Callback(object state)
             timers.Add(name, timer);
 
             // Now it's ok to start the timer.
-            timer.SafeTimerChange(dueTime, period, ref disposeEnded);
+            timer.SafeTimerChange(dueTime, period, ref _disposed);
         }
 
-        private bool disposeEnded;
+        private volatile bool _shuttingDown;
+        private bool _disposed;
+
         public void Dispose()
         {
-            if (disposeEnded)
-                return;
+            if (_disposed) return;
+            _shuttingDown = true;
 
             var timeout = TimeSpan.FromSeconds(1);
 
-            using (var waitHandle = new ManualResetEvent(false))
+            var keys = new List<string>(timers.Keys.AsEnumerable());
+
+            foreach (var k in keys)
+            {
+                timers[k]?.Change(Timeout.Infinite, Timeout.Infinite);
+            }
+
+            foreach (var k in keys)
             {
-                foreach (var task in timers)
+                var t = timers[k];
+                if (t is null) continue;
+
+                using (var done = new ManualResetEvent(false))
                 {
-                    // Returns false on second dispose
-                    if (task.Value.Dispose(waitHandle))
+                    var willSignal = t.Dispose(done);
+                    if (willSignal)
                     {
-                        if (!waitHandle.WaitOne(timeout))
-                        {
-                            throw new TimeoutException($"UNLEASH: Timeout waiting for task '{task.Key}' to stop..");
-                        }
+                        if (!done.WaitOne(timeout))
+                            throw new TimeoutException($"Timeout waiting for timer '{k}' to stop.");
                     }
+
+                    timers.Remove(k);
```

Comment:
Why this? We clear everything at the end don't we?

Reply:
Mmm fair enough. I'll remove this